commit 9ce89e4c3bd7e0809db2780f71144077d4fe1ca4
Author: Marc Zyngier <marc.zyngier@arm.com>
Date:   Tue May 2 14:30:40 2017 +0100

    KVM: arm/arm64: vgic-v3: Do not use Active+Pending state for a HW interrupt
    
    commit 3d6e77ad1489650afa20da92bb589c8778baa8da upstream.
    
    When an interrupt is injected with the HW bit set (indicating that
    deactivation should be propagated to the physical distributor),
    special care must be taken so that we never mark the corresponding
    LR with the Active+Pending state (as the pending state is kept in
    the physycal distributor).
    
    Fixes: 59529f69f504 ("KVM: arm/arm64: vgic-new: Add GICv3 world switch backend")
    Signed-off-by: Marc Zyngier <marc.zyngier@arm.com>
    Reviewed-by: Christoffer Dall <cdall@linaro.org>
    Signed-off-by: Christoffer Dall <cdall@linaro.org>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/virt/kvm/arm/vgic/vgic-v3.c b/virt/kvm/arm/vgic/vgic-v3.c
index e6b03fd..f132006 100644
--- a/virt/kvm/arm/vgic/vgic-v3.c
+++ b/virt/kvm/arm/vgic/vgic-v3.c
@@ -151,6 +151,13 @@ void vgic_v3_populate_lr(struct kvm_vcpu *vcpu, struct vgic_irq *irq, int lr)
 	if (irq->hw) {
 		val |= ICH_LR_HW;
 		val |= ((u64)irq->hwintid) << ICH_LR_PHYS_ID_SHIFT;
+		/*
+		 * Never set pending+active on a HW interrupt, as the
+		 * pending state is kept at the physical distributor
+		 * level.
+		 */
+		if (irq->active && irq->pending)
+			val &= ~ICH_LR_PENDING_BIT;
 	} else {
 		if (irq->config == VGIC_CONFIG_LEVEL)
 			val |= ICH_LR_EOI;
